searchCore/src/main/java/com/hamdan/forzenbook/search/core/viewmodel/BaseSearchViewModel.kt

```diff
@@ -13,50 +17,94 @@ abstract class BaseSearchViewModel(
 ) : ViewModel() {
 
     sealed interface SearchState {
-        val query: String
-
-        data class Searching(override val query: String = "") : SearchState
+        data class Searching(val query: String = "") : SearchState
+        object InvalidLogin : SearchState
     }
 
     protected abstract var searchState: SearchState
 
-    fun onNameClicked(id: Int, onSuccess: () -> Unit, onError: () -> Unit) =
-        searchById(id, onSuccess, onError)
+    fun onNameClicked(
+        id: Int,
+        context: Context,
+        onSuccess: () -> Unit,
+        onError: () -> Unit,
+    ) = searchById(id, context, onSuccess, onError)
 
-    fun onSearchSubmit(onSuccess: () -> Unit, onError: () -> Unit) =
-        searchByQuery(onSuccess, onError)
+    fun onSearchSubmit(
+        context: Context,
+        onSuccess: () -> Unit,
+        onError: () -> Unit,
+    ) = searchByQuery(context, onSuccess, onError)
 
     fun onUpdateSearch(text: String) = updateSearchText(text)
 
-    fun navigateQuery(): String = "/-1/${searchState.query}/"
+    fun navigateQuery(): String = "/-1/${searchState}/"
 
     fun navigateUser(id: Int): String = "/$id/%20/"
 
-    private fun searchById(id: Int, onSuccess: () -> Unit, onError: () -> Unit) {
-        viewModelScope.launch {
-            try {
-                searchForPostByIdUseCase(id)
-                onSuccess()
-            } catch (e: Exception) {
-                Log.v("Hamdan", e.message.toString())
-                onError()
-            }
+    private fun searchById(id: Int, context: Context, onSuccess: () -> Unit, onError: () -> Unit) {
+        try {
+            context.getSharedPreferences(
```

Comment:
Shared prefs access can block. Put in suspend function please

Reply:
ok, I didn't realize that, I'll put it in suspend